pkg/controller/kubelet-config/kubelet_config_bootstrap.go

```diff
@@ -0,0 +1,166 @@
+	"encoding/json"
+	"fmt"
+
+	ign3types "github.com/coreos/ignition/v2/config/v3_2/types"
+	"github.com/imdario/mergo"
+	mcfgv1 "github.com/openshift/machine-config-operator/pkg/apis/machineconfiguration.openshift.io/v1"
+	ctrlcommon "github.com/openshift/machine-config-operator/pkg/controller/common"
+	mtmpl "github.com/openshift/machine-config-operator/pkg/controller/template"
+	"github.com/openshift/machine-config-operator/pkg/version"
+	"github.com/vincent-petithory/dataurl"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kubeletconfigv1beta1 "k8s.io/kubelet/config/v1beta1"
+)
+
+// RunKubeletBootstrap generates MachineConfig objects for mcpPools that would have been generated by syncKubeletConfig
+// nolint:gocyclo
+func RunKubeletBootstrap(templateDir string, kubeletConfig *mcfgv1.KubeletConfig, controllerConfig *mcfgv1.ControllerConfig, mcpPools []*mcfgv1.MachineConfigPool) ([]*mcfgv1.MachineConfig, error) {
+	var (
+		kubeletIgnition            *ign3types.File
+		logLevelIgnition           *ign3types.File
+		autoSizingReservedIgnition *ign3types.File
+	)
+	userDefinedSystemReserved := make(map[string]string)
+
+	// Validate the KubeletConfig CR if exists
+	if err := validateUserKubeletConfig(kubeletConfig); err != nil {
+		return nil, newForgetError(err)
+	}
+
+	var res []*mcfgv1.MachineConfig
+	for _, pool := range mcpPools {
+		role := pool.Name
+		managedKey, err := getManagedKeyKubelet(pool, nil)
+		if err != nil {
+			return nil, err
+		}
+
+		// Generate the bootstrap KubeletConfig
+		originalKubeletIgn, err := generateBootstrapKubeletConfigs(controllerConfig, templateDir, role)
+		if err != nil {
+			return nil, err
+		}
+
+		if originalKubeletIgn.Contents.Source == nil {
+			return nil, err
+		}
+		dataURL, err := dataurl.DecodeString(*originalKubeletIgn.Contents.Source)
+		if err != nil {
+			return nil, err
+		}
+		originalKubeConfig, err := decodeKubeletConfig(dataURL.Data)
+		if err != nil {
+			return nil, err
+		}
+
+		if kubeletConfig != nil {
+			// Inject TLS Options from Spec
+			observedMinTLSVersion, observedCipherSuites := getSecurityProfileCiphers(kubeletConfig.Spec.TLSSecurityProfile)
+			originalKubeConfig.TLSMinVersion = observedMinTLSVersion
+			originalKubeConfig.TLSCipherSuites = observedCipherSuites
+		}
+
+		if kubeletConfig != nil && kubeletConfig.Spec.KubeletConfig != nil && kubeletConfig.Spec.KubeletConfig.Raw != nil {
+			specKubeletConfig, err := decodeKubeletConfig(kubeletConfig.Spec.KubeletConfig.Raw)
+			if err != nil {
+				return nil, err
+			}
+
+			if val, ok := specKubeletConfig.SystemReserved["memory"]; ok {
+				userDefinedSystemReserved["memory"] = val
+				delete(specKubeletConfig.SystemReserved, "memory")
+			}
+
+			if val, ok := specKubeletConfig.SystemReserved["cpu"]; ok {
+				userDefinedSystemReserved["cpu"] = val
+				delete(specKubeletConfig.SystemReserved, "cpu")
+			}
+
+			// Merge the Old and New
+			err = mergo.Merge(originalKubeConfig, specKubeletConfig, mergo.WithOverride)
```

Comment:
Similar to above, how do we make sure that this pool can take this kubeletconfig?

Kubeletconfigs unfortunately match based on labelling so it's much harder to parse the pool. The user has to define corresponding pool definitions as well to match label as well for this to work day 1 right?

Actually how does this work for containerruntimeconfigs?